Sofa/Component/Engine/Select/src/sofa/component/engine/select/BaseROI.h

```diff
@@ -0,0 +1,172 @@
+    Data<VecTetra> d_tetrahedraInROI; ///< Tetrahedra contained in the ROI
+    Data<VecHexa> d_hexahedraInROI; ///< Hexahedra contained in the ROI
+    Data< sofa::Size > d_nbIndices; ///< Number of selected indices
+
+    Data<VecCoord > d_pointsOutROI; ///< Points not contained in the ROI
+    Data<type::vector<Edge> > d_edgesOutROI; ///< Edges not contained in the ROI
+    Data<type::vector<Triangle> > d_trianglesOutROI; ///< Triangles not contained in the ROI
+    Data<type::vector<Quad> > d_quadsOutROI; ///< Quads not contained in the ROI
+    Data<type::vector<Tetra> > d_tetrahedraOutROI; ///< Tetrahedra not contained in the ROI
+    Data<type::vector<Hexa> > d_hexahedraOutROI; ///< Hexahedra not contained in the ROI
+    Data<SetIndex> d_indicesOut; ///< Indices of the points not contained in the ROI
+    Data<SetIndex> d_edgeOutIndices; ///< Indices of the edges not contained in the ROI
+    Data<SetIndex> d_triangleOutIndices; ///< Indices of the triangles not contained in the ROI
+    Data<SetIndex> d_quadOutIndices; ///< Indices of the quads not contained in the ROI
+    Data<SetIndex> d_tetrahedronOutIndices; ///< Indices of the tetrahedra not contained in the ROI
+    Data<SetIndex> d_hexahedronOutIndices; ///< Indices of the hexahedra not contained in the ROI
+
+    //Parameter
+    Data<bool> d_drawROI; ///< Draw the ROI. (default = false)
+    Data<bool> d_drawPoints; ///< Draw Points. (default = false)
+    Data<bool> d_drawEdges; ///< Draw Edges. (default = false)
+    Data<bool> d_drawTriangles; ///< Draw Triangles. (default = false)
+    Data<bool> d_drawQuads; ///< Draw Quads. (default = false)
+    Data<bool> d_drawTetrahedra; ///< Draw Tetrahedra. (default = false)
+    Data<bool> d_drawHexahedra; ///< Draw Tetrahedra. (default = false)
+    Data<double> d_drawSize; ///< rendering size for ROI and topological elements
+    Data<bool> d_doUpdate; ///< If true, updates the selection at the beginning of simulation steps. (default = true)
+protected:
+    BaseROI();
+    ~BaseROI() override = default;
+
+    bool isPointIn(const PointID pid);
+
+    template <typename Element>
+    bool isInROI(const Element & e);
+    
+    template <typename Element>
+    bool isInStrictROI(const Element & e);
+
+    // main function to implement from this interface
+    virtual bool isPointInROI(const CPos& p) = 0;
+
+    // special cases can be implemented by overriding those functions
+    virtual bool isEdgeInROI(const Edge& e);
+    virtual bool isEdgeInStrictROI(const Edge& e);
+    virtual bool isTriangleInROI(const Triangle& t);
+    virtual bool isTriangleInStrictROI(const Triangle& t);
+    virtual bool isQuadInROI(const Quad& q);
+    virtual bool isQuadInStrictROI(const Quad& q);
+    virtual bool isTetrahedronInROI(const Tetra& t);
+    virtual bool isTetrahedronInStrictROI(const Tetra& t);
+    virtual bool isHexahedronInROI(const Hexa& t);
+    virtual bool isHexahedronInStrictROI(const Hexa& t);
```

Comment:
```suggestion
    bool isPointIn(const PointID pid) const;

    template <typename Element>
    bool isInROI(const Element & e) const;
    
    template <typename Element>
    bool isInStrictROI(const Element & e) const;

    // main function to implement from this interface
    virtual bool isPointInROI(const CPos& p) const = 0;

    // special cases can be implemented by overriding those functions
    virtual bool isEdgeInROI(const Edge& e) const;
    virtual bool isEdgeInStrictROI(const Edge& e) const;
    virtual bool isTriangleInROI(const Triangle& t) const;
    virtual bool isTriangleInStrictROI(const Triangle& t) const;
    virtual bool isQuadInROI(const Quad& q) const;
    virtual bool isQuadInStrictROI(const Quad& q) const;
    virtual bool isTetrahedronInROI(const Tetra& t) const;
    virtual bool isTetrahedronInStrictROI(const Tetra& t) const;
    virtual bool isHexahedronInROI(const Hexa& t) const;
    virtual bool isHexahedronInStrictROI(const Hexa& t) const;
```
Given the name of the functions, I would expect that these functions are const.